psclient/src/main/java/com/majeur/psclient/service/ReplayManager.kt

```diff
@@ -0,0 +1,110 @@
+package com.majeur.psclient.service
+
+import com.majeur.psclient.service.observer.BattleRoomMessageObserver
+import com.majeur.psclient.ui.BaseFragment
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+import okhttp3.HttpUrl
+import okhttp3.OkHttpClient
+import okhttp3.Request
+import org.json.JSONObject
+
+class ReplayManager(
+        val httpClient: OkHttpClient,
+        val replayCallback: ReplayCallback) {
+
+    private val fragmentScope = BaseFragment.FragmentScope()
+    private lateinit var currentReplay : ReplayData
+
+
+    fun downloadAndStartReplay(uri : String) {
+        val url = HttpUrl.get(uri)
+        fragmentScope.launch {
+            val request = Request.Builder()
+                    .url(url)
+                    .build()
+            val replayResponse = withContext(Dispatchers.IO) {
+                val response = httpClient.newCall(request).execute()
+                response.body()?.string() ?: ""
+            }
+
+            currentReplay = ReplayData(JSONObject(replayResponse))
+
+            initReplayRoom(currentReplay)
```

Comment:
Maybe dispatch a |popup| message if replay cannot be retrieved.

Reply:
Done!